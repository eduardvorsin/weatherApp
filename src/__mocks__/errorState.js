import fetchingStatuses from '../redux/fetchingStatuses';

const errorData = {
  data: {},
  loadingStatus: fetchingStatuses.error,
  error: 'mock error',
};

const errorState = {
  weather: {
    hourlyWeather: errorData,
    currentWeather: errorData,
    dailyWeather: errorData,
    humidity: errorData,
    wind: errorData,
  },
  geocoding: errorData,
};

export default errorState;
