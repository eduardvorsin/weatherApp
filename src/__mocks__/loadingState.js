import fetchingStatuses from '../redux/fetchingStatuses';

const loadingData = {
  data: {},
  loadingStatus: fetchingStatuses.loading,
  error: null,
};

const loadingState = {
  weather: {
    hourlyWeather: loadingData,
    currentWeather: loadingData,
    dailyWeather: loadingData,
    humidity: loadingData,
    wind: loadingData,
  },
  geocoding: loadingData,
};

export default loadingState;
