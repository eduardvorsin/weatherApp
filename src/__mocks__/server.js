import { rest } from 'msw';
import { setupServer } from 'msw/node';
import mockGeocodingError from './geocodingError';
import mockWeatherError from './weatherError';

const isCurrentWeatherOrWindRequest = (searchParams) => (
  searchParams.get('current_weather') === 'true'
);

const isHourlyWeatherRequest = (searchParams) => {
  const currentParams = searchParams.getAll('hourly');
  return currentParams.includes('temperature_2m') && currentParams.includes('weathercode');
};

const isDailyWeatherRequest = (searchParams) => {
  const currentParams = searchParams.getAll('daily');
  return currentParams.includes('temperature_2m_max') && currentParams.includes('temperature_2m_min') && currentParams.includes('weathercode');
};

const isHumidityRequest = (searchParams) => {
  const currentParams = searchParams.getAll('hourly');
  return currentParams.includes('apparent_temperature') && currentParams.includes('relativehumidity_2m');
};

const currentWeatherData = {
  current_weather: {
    weathercode: 1,
    temperature: -6.7,
    winddirection: 128,
    windspeed: 1.64,
  },
};
const hourlyWeatherData = {
  hourly: {
    time: [
      '2022-12-27T03:00',
      '2022-12-27T04:00',
      '2022-12-27T05:00',
      '2022-12-27T06:00',
      '2022-12-27T07:00',
      '2022-12-27T08:00',
      '2022-12-27T09:00',
      '2022-12-27T10:00',
      '2022-12-27T11:00',
      '2022-12-27T12:00',
      '2022-12-27T13:00',
      '2022-12-27T14:00',
      '2022-12-27T15:00',
      '2022-12-27T16:00',
      '2022-12-27T17:00',
      '2022-12-27T18:00',
      '2022-12-27T19:00',
      '2022-12-27T20:00',
      '2022-12-27T21:00',
      '2022-12-27T22:00',
      '2022-12-27T23:00',
      '2022-12-28T00:00',
      '2022-12-28T01:00',
      '2022-12-28T02:00',
      '2022-12-28T03:00',
      '2022-12-28T04:00',
      '2022-12-28T05:00',
      '2022-12-28T06:00',
      '2022-12-28T07:00',
      '2022-12-28T08:00',
    ],
    temperature_2m: [
      -4.5, -4.4, -4.3, -4.5, -5.1, -5.6, -5.8, -6, -6.5, -6.8, -7, -6.4,
      -5.3, -4.3, -3.6, -3.4, -3.7, -4.4, -4.9, -5.3, -5.6, -5.8, -5.7, -5.5,
      -5.0, -4.5, -3.9, -3.1, -3.4, -4.9,
    ],
    weathercode: [1, 2, 3, 71, 71, 3, 3, 2, 2, 2, 1, 3, 71, 86, 1, 1, 3, 2, 2,
      71, 71, 3, 2, 1, 71, 2, 3, 1, 2, 71],
  },
};
const dailyWeatherData = {
  daily: {
    time: ['2022-12-26', '2022-12-27', '2022-12-29', '2022-12-30', '2022-12-31', '2023-01-01', '2023-01-02'],
    temperature_2m_max: [-9.4, -8, -7.3, -5, -6.7, -7.6, -9.8],
    temperature_2m_min: [-17.1, -15.5, -16.4, -17.5, -15.7, -16.3, -13.6],
    weathercode: [3, 2, 71, 2, 2, 3, 71],
  },
};
const humidityData = {
  hourly: {
    time: [
      '2022-12-27T00:00',
      '2022-12-27T01:00',
      '2022-12-27T02:00',
      '2022-12-27T03:00',
      '2022-12-27T04:00',
      '2022-12-27T05:00',
      '2022-12-27T06:00',
      '2022-12-27T07:00',
      '2022-12-27T08:00',
      '2022-12-27T09:00',
    ],
    apparent_temperature: [-17.4, -18, -18.6, -19.1, -19.5, -19, -17, -16, -15, 14],
    relativehumidity_2m: [81, 82, 83, 84, 85, 86, 71, 72, 73, 74],
  },
};
const geocodingData = {
  address: {
    city: 'Вестминстер',
    state_district: 'Большой Лондон',
    state: 'Англия',
    country: 'Великобритания',
    country_code: 'gb',
    'ISO3166-2-lvl4': 'GB-ENG',
    'ISO3166-2-lvl8': 'GB-WSM',
  },
};

const handlers = [
  rest.get('https://api.open-meteo.com/v1/forecast', (req, res, ctx) => {
    const { searchParams } = req.url;

    if (isCurrentWeatherOrWindRequest(searchParams)) {
      return res(ctx.json(currentWeatherData), ctx.delay(150));
    }
    if (isHourlyWeatherRequest(searchParams)) {
      return res(ctx.json(hourlyWeatherData), ctx.delay(150));
    }
    if (isDailyWeatherRequest(searchParams)) {
      return res(ctx.json(dailyWeatherData), ctx.delay(150));
    }
    if (isHumidityRequest(searchParams)) {
      return res(ctx.json(humidityData), ctx.delay(150));
    }

    return res(ctx.json('unknown request'), ctx.delay(150));
  }),
  rest.get('https://nominatim.openstreetmap.org/reverse', (_, res, ctx) => (
    res(ctx.json(geocodingData), ctx.delay(150))
  )),
];

export const errorHandlers = [
  rest.get('https://nominatim.openstreetmap.org/reverse', (_, res, ctx) => res(ctx.status(404), ctx.json(mockGeocodingError), ctx.delay(150))),
  rest.get('https://api.open-meteo.com/v1/forecast', (_, res, ctx) => res(ctx.status(404), ctx.json(mockWeatherError), ctx.delay(150))),
];

const server = setupServer(...handlers);

export const setupServerWithErrors = () => server.use(...errorHandlers);

export default server;
