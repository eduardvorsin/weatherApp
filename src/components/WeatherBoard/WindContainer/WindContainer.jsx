import PropTypes from 'prop-types';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import fetchingStatuses from '../../../redux/fetchingStatuses';
import selectWind from '../../../redux/selectors/selectWind/selectWind';
import fetchWind from '../../../redux/thunks/fetchWind/fetchWind';
import ErrorInfo from '../../UI/ErrorInfo/ErrorInfo';
import Spinner from '../../UI/Spinner/Spinner';
import Wind from '../Wind/Wind';

export default function WindContainer({
  className,
  location,
  testId,
}) {
  const dispatch = useDispatch();
  const {
    data,
    loadingStatus,
  } = useSelector(selectWind);

  useEffect(() => {
    if (location.latitude && location.longitude) {
      dispatch(fetchWind(location));
    }
  }, [location, dispatch]);

  return (
    <div
      className={`wind__section wind-section ${className}`}
      data-testid={testId}
    >
      {loadingStatus === fetchingStatuses.loading && (
        <Spinner
          size={246}
        />
      )}
      {loadingStatus === fetchingStatuses.error && (
        <ErrorInfo
          message="Не удалось загрузить данные о ветре"
          headingLevel={3}
        />
      )}
      {loadingStatus === fetchingStatuses.resolved
        && (
          <Wind
            windSpeed={data.speed}
            windDirectionAngle={data.directionAngle}
          />
        )}
    </div>
  );
}

WindContainer.defaultProps = {
  className: '',
  location: {},
  testId: '',
};

WindContainer.propTypes = {
  className: PropTypes.string,
  testId: PropTypes.string,
  location: PropTypes.shape({
    latitude: PropTypes.number,
    longitude: PropTypes.number,
  }),
};
