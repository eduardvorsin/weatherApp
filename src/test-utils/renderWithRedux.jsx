import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createReduxStore } from '../redux/store';

export default function renderWithRedux(component, initialState) {
  const store = createReduxStore(initialState);

  return render(
    <Provider store={store}>
      {component}
    </Provider>,
  );
}
