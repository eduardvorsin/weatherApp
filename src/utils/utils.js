export const isCorrectHeadingLevel = (level) => {
  if (Number.isNaN(+level)) {
    throw new Error('the heading level must be a number');
  }

  if (level < 1 || level > 6) {
    throw new Error('the heading level must be at least 1 and not more than 6');
  }

  if (!Number.isInteger(level)) {
    throw new Error('the header level must be an integer');
  }

  return true;
};

export const getWeatherStatusByCode = (code) => {
  if (typeof code !== 'number') {
    throw new Error('code parameter should be a number type');
  }

  let status = '';

  if (code === 0 || code === 1) {
    status = 'ясно';
  } else if (code === 2 || code === 3) {
    status = 'пасмурно';
  } else if (code === 45) {
    status = 'туман';
  } else if ((code >= 48 && code <= 67) || (code >= 80 && code <= 86)) {
    status = 'дождь';
  } else if (code >= 71 && code <= 77) {
    status = 'снег';
  } else if (code >= 95 && code <= 99) {
    status = 'гроза';
  } else {
    throw new Error('recived unknown weather code');
  }

  return status;
};

export const getIconPathByCode = (code, basePath = '/assets/icons') => {
  if (typeof basePath !== 'string') {
    throw new Error('basePath parameter should be a string type');
  }

  if (typeof code !== 'number') {
    throw new Error('code parameter should be a number type');
  }

  let path = '';

  if (code === 0 || code === 1) {
    path += 'sunny';
  } else if (code === 2 || code === 3) {
    path += 'cloudy';
  } else if (code === 45) {
    path += 'fog';
  } else if ((code >= 48 && code <= 67) || (code >= 80 && code <= 86)) {
    path += 'rain';
  } else if (code >= 71 && code <= 77) {
    path += 'snow';
  } else if (code >= 95 && code <= 99) {
    path += 'thunderstorm';
  } else {
    throw new Error('recived unknown weather code');
  }

  return `${basePath}/${path}.svg`;
};

export const getFullDayDescription = (date) => {
  if (Number.isNaN(new Date(date).getTime())) {
    throw new Error('invalid date string passed');
  }

  const options = {
    weekday: 'long',
    day: 'numeric',
    month: 'long',
  };

  return new Intl.DateTimeFormat('ru-RU', options).format(new Date(date));
};
