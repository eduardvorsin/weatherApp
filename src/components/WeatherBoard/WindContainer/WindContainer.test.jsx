import { screen, waitForElementToBeRemoved } from '@testing-library/react';
import renderWithRedux from '../../../test-utils/renderWithRedux';
import { getWindDirectionNameByAngle } from '../../../utils/utils';
import errorState from '../../../__mocks__/errorState';
import loadingState from '../../../__mocks__/loadingState';
import mockLocation from '../../../__mocks__/location';
import resolvedState from '../../../__mocks__/resolvedState';
import server, { setupServerWithErrors } from '../../../__mocks__/server';
import WindContainer from './WindContainer';

describe('WindContainer component tests', () => {
  it('renders correctly', () => {
    renderWithRedux(<WindContainer
      className="test-class"
      location={{}}
      testId="wind"
    />);

    expect(screen.getByTestId('wind')).toBeInTheDocument();
  });

  it('renders with error when loadingStatus equals "error"', () => {
    renderWithRedux(
      <WindContainer
        className="test-class"
        location={{}}
      />,
      errorState,
    );

    expect(screen.getByText(/Не удалось загрузить данные о ветре/i)).toBeInTheDocument();
  });

  it('renders with loading spinner when loadingStatus equals "loading"', () => {
    renderWithRedux(
      <WindContainer
        className="test-class"
        location={{}}
      />,
      loadingState,
    );

    expect(screen.getByTitle('loading spinner')).toBeInTheDocument();
  });

  it('renders with data when loadingStatus equals "resolved"', () => {
    renderWithRedux(
      <WindContainer
        className="test-class"
        location={{}}
      />,
      resolvedState,
    );

    const directionAngle = new RegExp(getWindDirectionNameByAngle(
      resolvedState.weather.wind.data.directionAngle,
    ));

    const speed = new RegExp(
      resolvedState.weather.wind.data.speed,
    );

    expect(screen.getByText(speed)).toBeInTheDocument();
    expect(screen.getByText(directionAngle)).toBeInTheDocument();
  });

  it('the classname is correctly assigned to the element', () => {
    renderWithRedux(
      <WindContainer
        className="test-class"
        location={{}}
        testId="wind"
      />,
    );

    expect(screen.getByTestId('wind')).toHaveClass('test-class');
  });

  it('snapshot without anything', () => {
    renderWithRedux(
      <WindContainer
        className="test-class"
        location={{}}
        testId="wind"
      />,
    );
    expect(screen.getByTestId('wind')).toMatchSnapshot();
  });

  it('snapshot with mock error', () => {
    renderWithRedux(
      <WindContainer
        className="test-class"
        location={{}}
        testId="wind"
      />,
      errorState,
    );

    expect(screen.getByTestId('wind')).toMatchSnapshot();
  });

  it('snapshot with mock data', () => {
    renderWithRedux(
      <WindContainer
        className="test-class"
        location={{}}
        testId="wind"
      />,
      resolvedState,
    );

    expect(screen.getByTestId('wind')).toMatchSnapshot();
  });

  it('snapshot with a spinner', () => {
    renderWithRedux(
      <WindContainer
        className="test-class"
        location={{}}
        testId="wind"
      />,
      loadingState,
    );

    expect(screen.getByTestId('wind')).toMatchSnapshot();
  });
});

describe('WindContainer component itegration tests', () => {
  beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it('the loading status changes to the error status', async () => {
    setupServerWithErrors();

    renderWithRedux(
      <WindContainer
        location={mockLocation}
      />,
    );

    expect(screen.getByTitle('loading spinner')).toBeInTheDocument();
    expect(screen.queryByText(/ветер/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/скорость/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/направление/i)).not.toBeInTheDocument();

    await waitForElementToBeRemoved(screen.getAllByTitle('loading spinner'));

    expect(screen.queryByTitle('loading spinner')).not.toBeInTheDocument();
    expect(screen.queryByText(/ветер/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Не удалось загрузить данные о ветре/i)).toBeInTheDocument();
  });

  it('the loading status changes to the resolved status', async () => {
    renderWithRedux(
      <WindContainer
        location={mockLocation}
      />,
    );

    expect(screen.getByTitle('loading spinner')).toBeInTheDocument();
    expect(screen.queryByText(/ветер/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/скорость/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/направление/i)).not.toBeInTheDocument();

    await waitForElementToBeRemoved(screen.getAllByTitle('loading spinner'));

    expect(screen.queryByTitle('loading spinner')).not.toBeInTheDocument();
    expect(screen.queryByText(/ветер/i)).toBeInTheDocument();
    expect(screen.queryByText(/Не удалось загрузить данные о ветре/i)).not.toBeInTheDocument();
  });
});
