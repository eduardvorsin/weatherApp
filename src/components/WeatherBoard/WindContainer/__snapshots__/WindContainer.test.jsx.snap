// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WindContainer component tests snapshot with a spinner 1`] = `
<div
  class="wind__section wind-section test-class"
  data-testid="wind"
>
  <div
    aria-label="loading spinner"
    class="spinner "
    style="--size: 246px;"
    title="loading spinner"
  >
    <div
      class="spinner__ring"
      style="--ring-size: 196.8px;"
    />
    <div
      class="spinner__ring"
      style="--ring-size: 196.8px;"
    />
    <div
      class="spinner__ring"
      style="--ring-size: 196.8px;"
    />
    <div
      class="spinner__ring"
      style="--ring-size: 196.8px;"
    />
  </div>
</div>
`;

exports[`WindContainer component tests snapshot with mock data 1`] = `
<div
  class="wind__section wind-section test-class"
  data-testid="wind"
>
  <div
    class="wind "
    data-testid=""
  >
    <h3
      class="wind__title"
    >
      Ветер
    </h3>
    <div
      class="wind__content"
    >
      <svg
        class="wind__icon"
        fill="none"
        viewBox=" 0 0 850 850"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M464.387 409.738c7.981-9.277 12.804-21.348 12.804-34.546 0-29.271-23.729-53-53-53s-53 23.729-53 53a52.769 52.769 0 0 0 10.626 31.84l-28.143 31.105-1.441 1.593-.208 2.138-29.964 307.63h-34.567v13h274.5v-13h-34.066l-29.964-307.63-.209-2.138-1.441-1.593-28.5-31.5-3.427 3.101Zm-5.953 5.909a52.787 52.787 0 0 1-34.243 12.545 52.781 52.781 0 0 1-33.325-11.785l-26.111 28.859-29.633 304.232h179.744l-29.633-304.232-26.799-29.619Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
        <path
          class="wind__blades"
          clip-rule="evenodd"
          d="m710.884 169.192-4.597 4.596-4 4-146.999 147-1.5 1.5-4.597 4.596-4.596-4.596-72-72-4.596-4.596 4.596-4.596 2-2 118-118 32.5-32.5L629.691 88l4.597 4.596 12.499 12.5 59.5 59.5 4.597 4.596Zm-18.385 0-54.904-54.903-7.904-7.904-27.903 27.903-115.404 115.404 62.807 62.808 143.308-143.308ZM344.788 462.788l.193-.193.5.5 79.211-79.211 79.903 79.904.25-.25.053.053 9.193-9.193-79.957-79.956 79.154-79.154-9.193-9.192-.807.807-.5-.5-78.346 78.346-78.154-78.153-9.192 9.192 2.057 2.057-.25.25 76.096 76.097-79.403 79.403 9.192 9.193ZM138 579.691l4.596-4.596 4-4 147-146.999 1.5-1.5 4.596-4.597 4.596 4.597 72 71.999 4.596 4.597-4.596 4.596-2 2-118 117.999-32.5 32.5-4.596 4.597-4.596-4.597-12.5-12.5-59.5-59.499-4.596-4.597Zm18.385 0 54.903 54.904 7.904 7.904 27.904-27.904L362.5 499.192l-62.808-62.808-143.307 143.307ZM220.192 89l4.596 4.596 4 4 147 147 1.5 1.5 4.596 4.596-4.596 4.596-72 72-4.596 4.596-4.596-4.596-2-2-118-118-32.5-32.5-4.596-4.596 4.596-4.596 12.5-12.5 59.5-59.5L220.192 89Zm0 18.385-54.903 54.903-7.904 7.904 27.903 27.904L300.692 313.5l62.808-62.808-143.308-143.307Zm410.499 553.999-4.596-4.597-4-4-147-146.999-1.5-1.5-4.596-4.596 4.596-4.597 72-71.999 4.596-4.597 4.597 4.597 2 2 117.999 117.999 32.5 32.5 4.597 4.596-4.597 4.597-12.5 12.5-59.499 59.499-4.597 4.597Zm0-18.385 54.904-54.904 7.904-7.904-27.904-27.903-115.404-115.404-62.807 62.808 143.307 143.307Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <div
        class="wind__text"
      >
        <p
          class="wind__speed"
        >
          Скорость:
           
          1.64
          м/с
        </p>
        <p
          class="wind__direction"
        >
          Направление:
           
          Юго-восток
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`WindContainer component tests snapshot with mock error 1`] = `
<div
  class="wind__section wind-section test-class"
  data-testid="wind"
>
  <div
    class="error-info  "
    data-testid=""
    role="alert"
  >
    <h3
      class="error-info__description"
    >
      Не удалось загрузить данные о ветре
    </h3>
  </div>
</div>
`;

exports[`WindContainer component tests snapshot without anything 1`] = `
<div
  class="wind__section wind-section test-class"
  data-testid="wind"
/>
`;
